pkgs/shells/fish/default.nix

```diff
@@ -16,11 +16,18 @@
 , ncurses
 , python3
 , cmake
+, fishPlugins
 
 , runCommand
 , writeText
 , nixosTests
 , useOperatingSystemEtc ? true
+  # A path to a file to source early during initialization to set up the environment.
+  # In particular, if this file defines NIX_PROFILES then fish can set up its function,
+  # completion, and conf.d paths accordingly. If set to an attrset with `bash = true` the
+  # file will be interpreted as a bash file and parsed with fishPackages.foreign-env.
+  # Note: This path is not copied into the nix store but is referenced directly.
+, fishEnvPreInit ? { path = "/etc/fish/nixos-env-preinit.fish"; bash = false; }
```

Comment:
Why not just make this basically `with types; nullOr lines` and let the user handle everything they need to? Seems kinda restrictive if you can only choose one path, rather than being able to do whatever they want in the preinit (e.g. setting up `fenv` and then using it to source `"${nix}/etc/profile.d/nix-daemon.sh"` as in your example).

Basically, I'm imagining:

```nix
# [snip]
, fishEnvPreInit ? null
}:

# [snip]

${lib.optionalString (fishEnvPreInit != null) ''
begin
  ${fishEnvPreinit}
end
''}
```

which would be much more flexible, IMO.

Reply:
I was originally going to do it that way, but I decided it was a bit onerous for people to handle `fenv` themselves. I suppose another option is to do something like

```nix
  # An optional string containing Fish code that initializes the environment.
  # This is run at the very beginning of initialization. If it sets $NIX_PROFILE
  # then Fish will use that configure its function, completion, and conf.d paths.
  # For example:
  #   fishEnvPreInit = "source /etc/fish/my-env-preinit.fish";
  # It can also be a function that takes one argument, which is a function that
  # takes a path to a bash file and converts it to fish. For example:
  #   fishEnvPreInit = source: source "${nix}/etc/profile.d/nix-daemon.sh";
, fishEnvPreInit ? null
```

This way I can still hide the fenv complexity from the user.

Reply:
Sounds like an OK compromise to me. I'd just make sure it's obvious that the conversion uses `fishPlugins.foreign-env` and not, say, `babelfish`.

Reply:
I'm not sure it really matters to the user _how_ the sourcing works, only that it can source bash files this way.